tests/gem/test_packets.py

```diff
@@ -144,6 +145,52 @@ def test_packet_delta_polarized_watt_seconds(self):
         )
 
 
+class TestPacketAverageComputation(unittest.TestCase):
+    def test_packet_average_power(self):
+        packet_maker = functools.partial(
+            packets.Packet,
+            packet_format=packets.BIN32_ABS,
+            voltage=120.0,
+            device_id=123456,
+            serial_number=123456,
+            pulse_counts=[0] * packets.PacketFormat.NUM_PULSE_COUNTERS,
+            temperatures=[0] * packets.PacketFormat.NUM_TEMPERATURE_SENSORS,
+        )
+        packet_a = packet_maker(
+            seconds=0,
+            absolute_watt_seconds=[10] * packets.BIN32_ABS.num_channels,
+        )
+        packet_b = packet_maker(
+            seconds=10,
+            absolute_watt_seconds=[20] * packets.BIN32_ABS.num_channels,
+        )
+        self.assertEqual(packet_a.get_average_power(0, packet_b), 1.0)
+        self.assertEqual(packet_b.get_average_power(0, packet_a), 1.0)
+
+    def test_packet_average_power_net_metering(self):
+        packet_maker = functools.partial(
+            packets.Packet,
+            packet_format=packets.BIN32_NET,
+            voltage=120.0,
+            device_id=123456,
+            serial_number=123456,
+            pulse_counts=[0] * packets.PacketFormat.NUM_PULSE_COUNTERS,
+            temperatures=[0] * packets.PacketFormat.NUM_TEMPERATURE_SENSORS,
+        )
+        packet_a = packet_maker(
+            seconds=0,
+            absolute_watt_seconds=[10] * packets.BIN32_NET.num_channels,
+            polarized_watt_seconds=[0] * packets.BIN32_NET.num_channels,
+        )
+        packet_b = packet_maker(
+            seconds=10,
+            absolute_watt_seconds=[40] * packets.BIN32_NET.num_channels,
+            polarized_watt_seconds=[10] * packets.BIN32_NET.num_channels,
+        )
+        self.assertEqual(packet_a.get_average_power(0, packet_b), 1.0)
+        self.assertEqual(packet_b.get_average_power(0, packet_a), 1.0)
```

Comment:
Best I can tell, the logic is copied correctly from [the original code](https://github.com/jkeljo/greeneye-monitor/blob/master/greeneye/monitor.py#L217), but I don't think this test result is right.  At least, I can't convince myself it's right; I think the average power here should be 2, not 1

Reply:
@jkeljo, any thoughts here?

Reply:
Per @benk22 [on Brultech's forum](https://www.brultech.com/community/viewtopic.php?p=14655#p14655):
> The GEM has 2 internal counters for each channel, absolute wattseconds and polarized wattseconds. Polarized is outgoing power, absolute is incoming and outgoing combined.

To illustrate what the test is doing, here are our values:

| Packet | Polarized (W) | Absolute (W) |
| --- | --- | --- |
| A | 0 | 10 |
| B | 10 | 40 |

Here are the delta, given that information:

| Delta | Value |
| --- | --- |
| Polarized | 10 W |
| Absolute | 30 W |
| Produced | 10 W |
| Consumed | 20 W (absolute - polarized) |

The code, taken from jkeljo/greeneye and copied in this PR, gets an average power from the upstream power source (probably the utility company).  I don't think that's the most useful metric, however, especially in a home automation context.  The total kWh are useful for understanding billing information, but in a given moment, I'm more interested in how much average power I'm producing and consuming.  In fact, if I want to try and create some smarts to figure out what devices are active, I need to know what the consumption is in the home, as opposed to what we're consuming from the upstream power source.  My previous confusion was that this was trying to get the value of average consumed power, hence thinking it should be 2.

Therefore, what I actually want to do with this PR is to get those two values (produced and consumed) independently.  That means this isn't as useful for the downstream library that has this code now, but maybe you'll want to change what that does and have both values split out as well.

Reply:
I got confused by point of measurement (which is what we have) and overall system math.  It's safe to fully ignore that last comment of mine 😅

There was a bug in the code I had written, however, which was caught in a new test I wrote.  I also added a big comment explaining how this works so hopefully nobody else gets confused by this like I did.